test_util/test_upgrade_vpc.py

```diff
@@ -82,39 +119,97 @@ def get_test_app():
                 "path": "/",
                 "portIndex": 0,
                 "gracePeriodSeconds": 5,
-                "intervalSeconds": 10,
-                "timeoutSeconds": 10,
-                "maxConsecutiveFailures": 3
+                "intervalSeconds": 1,
+                "timeoutSeconds": 5,
+                "maxConsecutiveFailures": 1
             }
         ],
     }
-    return app
 
+    # DNS resolution app to make sure DNS is available during the upgrade.
+    # Periodically resolves the healthcheck app's domain name and logs whether
+    # it succeeded to a file in the Mesos sandbox.
+    dns_app = {
+        "id": '/' + TEST_APP_NAME_FMT.format('dns-' + test_id),
+        "cmd": """
+while true
+do
+    printf "%s " $(date --utc -Iseconds) >> $MESOS_SANDBOX/$DNS_LOG_FILENAME
+    if host -W $TIMEOUT_SECONDS $RESOLVE_NAME
+    then
+        echo SUCCESS >> $MESOS_SANDBOX/$DNS_LOG_FILENAME
+    else
+        echo FAILURE >> $MESOS_SANDBOX/$DNS_LOG_FILENAME
+    fi
+    sleep $INTERVAL_SECONDS
+done
+""",
+        "env": {
+            'RESOLVE_NAME': marathon_app_id_to_mesos_dns_name(healthcheck_app['id']) + '.marathon.mesos',
+            'DNS_LOG_FILENAME': 'dns_resolve_log.txt',
+            'INTERVAL_SECONDS': '1',
+            'TIMEOUT_SECONDS': '1',
+        },
+        "cpus": 0.5,
+        "mem": 32.0,
+        "instances": 1,
+        "container": {
+            "type": "DOCKER",
+            "docker": {
+                "image": "mapitman/bind-utils",
+                "network": "BRIDGE",
+            }
+        },
+        "dependencies": [healthcheck_app['id']],
+    }
+
+    # Deploy test apps.
+    # TODO(branden): Define apps and deploy_app() such that we can deploy all
+    # test apps at once and then block on waiting for them all to finish
+    # deploying. Marathon ought to be responsible for deploying apps in the
+    # right order and knowing whether they're deployed and stable.
+    cluster_api.marathon.deploy_app(healthcheck_app)
+    cluster_api.marathon.ensure_deployments_complete()
+    # TODO(branden): This is a hack to make sure we don't deploy dns_app before
```

Comment:
It's a symptom of the problem described in the TODO above it. The TODO here implies that I'd like to remove the hack. I'll create a ticket for this and link it here like you suggested above.